--- brainbash	2019-04-04 16:49:47.758549174 +0800
+++ brainbash3	2019-04-04 16:50:48.469553537 +0800
@@ -6,19 +6,8 @@
 	exit 1
 }
 
-if (( $# )); then
-	if [[ -e $1 ]]; then
-		program=$(< "$1")
-	else
-		die "file '$1' does not exist"
-	fi
-else
-	mapfile -t
-	program=${MAPFILE[*]}
-fi
-
+program=$(< "${1:-/dev/stdin}") || exit
 program=${program//[^'><+-.,[]']}
-declare -A tape
 
 for (( i = ptr = 0; i < ${#program}; ++i )); do
 	case ${program:i:1} in
@@ -26,18 +15,18 @@
 			(( ++ptr ))
 			;;
 		'<')
-			(( --ptr ))
+			(( ptr && --ptr ))
 			;;
 		'+')
-			(( ++tape['$ptr'] ))
-			(( tape['$ptr'] > 255 )) && tape[$ptr]=0
+			(( ++tape[ptr] ))
+			(( tape[ptr] > 255 )) && tape[ptr]=0
 			;;
 		'-')
-			(( --tape['$ptr'] ))
-			(( tape['$ptr'] < 0 )) && tape[$ptr]=255
+			(( --tape[ptr] ))
+			(( tape[ptr] < 0 )) && tape[ptr]=255
 			;;
 		'.')
-			printf -v f %x "${tape[$ptr]}"
+			printf -v f %x "${tape[ptr]}"
 			printf %b "\x$f"
 			;;
 		',')
@@ -45,12 +34,12 @@
 
 			[[ $REPLY ]] && {
 				printf -v f %d "'${REPLY::1}"
-				tape[$ptr]=$f
+				tape[ptr]=$f
 				REPLY=${REPLY:1}
 			}
 			;;
 		'[')
-			if (( tape['$ptr'] )); then
+			if (( tape[ptr] )); then
 				stack+=("$i")
 			else
 				for (( depth = 1; depth > 0 && ++i; )); do
@@ -65,7 +54,7 @@
 		']')
 			(( _ = ${#stack[@]} )) || die "unmatched ]"
 
-			(( tape['$ptr'] )) && (( i = stack[_-1] - 1 ))
+			(( tape[ptr] )) && (( i = stack[_-1] - 1 ))
 			unset 'stack[_-1]'
 			;;
 	esac
